hvplot/plotting/scatter_matrix.py

```diff
@@ -38,19 +46,82 @@ def scatter_matrix(data, c=None, chart='scatter', diagonal='hist', alpha=0.5, **
+        if 'how' in kwds:
+            dn_kwds['how'] = kwds.pop('how')
+        if 'mask' in kwds:
+            dn_kwds['mask'] = kwds.pop('mask')
+
+    if spread:
+        sp_kwds = {}
+        if 'px' in kwds:
+            sp_kwds['px'] = kwds.pop('px')
+        if 'shape' in kwds:
+            sp_kwds['shape'] = kwds.pop('shape')
+        if 'how' in kwds:
+            sp_kwds['how'] = kwds.pop('how')
+        if 'mask' in kwds:
+            sp_kwds['mask'] = kwds.pop('mask')
+        
+
     colors = _hv.plotting.util.process_cmap('Category10', categorical=True)
     chart_opts = dict(alpha=alpha, cmap=colors, tools=['box_select', 'lasso_select'],
                       nonselection_alpha=0.1, **kwds)
 
+    #get initial scatter matrix.  No color.
     grid = _hv.operation.gridmatrix(data, diagonal_type=diagonal, chart_type=chart)
+
     if c:
+        #change colors for scatter matrix
         chart_opts['color_index'] = c
+        # not entirely sure what the next line is doing, but its helping
+        # with the coloring.
         grid = grid.map(lambda x: x.clone(vdims=x.vdims+[c]), 'Scatter')
+        # create a new scatter matrix with groups for each catetory, so now the histogram will
+        # show separate colors for each group.
         groups = _hv.operation.gridmatrix(data.groupby(c).overlay(),
                                           chart_type=chart,
                                           diagonal_type=diagonal)
+        # take the correct layer from each Overlay object within the scatter matrix.
         grid = (grid * groups).map(lambda x: x.get(0) if isinstance(x.get(0), chart) else x.get(1),
                                    _hv.Overlay)
 
+    # set the histogram colors 
     diagonal_opts = {'fill_color': _hv.Cycle(values=colors)}
-    return grid.options({chart.__name__: chart_opts, diagonal.__name__: diagonal_opts})
+    # actually changing to the same color scheme for both scatter and histogram plots.
+    grid = grid.options({chart.__name__: chart_opts, diagonal.__name__: diagonal_opts})
+    
+    # Perform datashade options after all the coloring is finished.
+    if datashade:
+        if dynspread:
+            grid = grid.map(lambda x: x.apply(hd.datashade, **ds_kwds)
+                                        .apply(hd.dynspread, **dn_kwds)
+                                        if isinstance(x, chart) else x)
+        elif spread:
+            grid = grid.map(lambda x: x.apply(hd.datashade, **ds_kwds)
+                                        .apply(hd.spread, **sp_kwds)
+                                        if isinstance(x, chart) else x)
+        else:
+            grid = grid.map(lambda x: x.apply(hd.datashade, **ds_kwds) 
+                            if isinstance(x, chart) else x)
+
+    elif rasterize:
+        grid = grid.map(lambda x: x.apply(hd.rasterize, **ds_kwds)
+                                        if isinstance(x, chart) else x)
```

Comment:
I ended up removing the `apply_to_charts` function, because you need to apply the options (dynspread or spread) for datashade right after performing the datashade function. If you try to run `apply_to_charts` once, and then do it a second time to apply dynspread or spread, it will check for elements that are of the type `scatter`.  These are no longer there after they were datashaded, so it doesn't do the second step. For some reason, trying to change things around to check for whether an element is of the type "hist" (or "histogram") doesn't work.  I was trying to keep the "apply_to_charts" function, by doing something like this:
```
def apply_to_charts(grid, op, kwds):
        return grid.map(lambda x: x
                        if isinstance(x, diagonal)
                        else x.apply(op, **kwds))```
or even:
```
def apply_to_charts(grid, op, kwds):
        return grid.map(lambda x: x.apply(op, **kwds)
                        if not isinstance(x, diagonal)
                        else x)```
where diagonal is the type for the diagonal elements.
This would give me an error that the histogram type did not any 'bounds'.